{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1d7b619-0938-476a-8132-ab2fd32a4016",
   "metadata": {},
   "source": [
    "# My first notebook\n",
    "\n",
    "> Note: to avoid random Lorem ipsum text, we've taken the step of asking chatGPT to explain the tools. It is therefore possible that there may be some oddities.\n",
    "\n",
    "## Our pipeline\n",
    "\n",
    "1. Download data\n",
    "2. Create a genome index\n",
    "3. Check quality\n",
    "4. Align data with reference genome\n",
    "5. Count number of read per gene\n",
    "6. Diferential analysis\n",
    "7. Visualisation \n",
    "\n",
    "## Load tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "101c9a32-2b41-4473-b4f3-70a2ee2fe55d",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load zenodo_get/1.3.2 \n",
    "module load hisat2/2.2.1\n",
    "module load fastqc/0.12.1\n",
    "module load samtools/1.15.1\n",
    "module load htseq/0.13.5\n",
    "module load r/4.2.3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f46bf8a-a2c3-471b-9492-9a7d46ef99ff",
   "metadata": {},
   "source": [
    "## Get Data from Zenodo\n",
    "\n",
    "Zenodo is an online platform that enables researchers and scientists to deposit, share, and archive their research data, publications, software, presentations, and other digital content. It is an initiative developed by the European Organization for Nuclear Research (CERN) to promote the free circulation of scientific knowledge and facilitate access to research findings.\n",
    "\n",
    "Zenodo provides free and permanent storage space for researchers, ensuring the longevity and availability of research data in the long term. Additionally, the platform assigns Digital Object Identifiers (DOIs) to each deposit, making it easier to cite and reference research works, thus contributing to the academic recognition of authors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1e2ea42-6a29-4c44-8f9e-fb33a530e599",
   "metadata": {},
   "source": [
    "### Get data from Zenodo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c68fb590-0276-493c-ad92-d931052a4eee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Title: reduced RNAseq for FAIR_Bioinfo courses\n",
      "Keywords: \n",
      "Publication date: 2020-08-24\n",
      "DOI: 10.5281/zenodo.3997237\n",
      "Total size: 118.8 MB\n",
      "\n",
      "Link: https://zenodo.org/api/records/3997237/files/FAIR_Bioinfo_data.tar.gz/content   size: 118.8 MB\n",
      " 19% [.........                                         ]  24502272 / 124534911\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub message rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_msg_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_msg_rate_limit=1000.0 (msgs/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checksum is correct. (85db07129cd29f353fed6c547f5ed4b1) ] 123969536 / 124534911\n",
      "All files have been downloaded.\n"
     ]
    }
   ],
   "source": [
    "zenodo_get -t 60 -R 3 10.5281/zenodo.3997237"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b1e75a0-3d53-4329-9b4a-aea4e9610558",
   "metadata": {},
   "source": [
    "### Checking the viability of data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2d5b4526-888e-4202-9907-817b1133de24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAIR_Bioinfo_data.tar.gz: OK\n"
     ]
    }
   ],
   "source": [
    "md5sum -c md5sums.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88badfb8-8448-4f3f-af0d-2dcd6c574070",
   "metadata": {},
   "source": [
    "### Extract data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8dedd100-796a-40ee-a128-1bd141be70bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tar -xzf FAIR_Bioinfo_data.tar.gz && rm FAIR_Bioinfo_data.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b69d853b-aa3e-4ce6-aa79-d519b6d5bfeb",
   "metadata": {},
   "source": [
    "## Create hisat2 indexes\n",
    "\n",
    "### The tool used : `hisat2-build`\n",
    "\n",
    "The `hisat2-build` command is used in bioinformatics to index a reference genome for efficient alignment of sequence reads. Here's what it does:\n",
    "\n",
    "1. **Reference Genome Indexing:** Takes a reference genome, divides it into small parts, and creates an index.\n",
    "   \n",
    "2. **Partitioning into Small Parts:** Divides the reference genome into fragments of a certain size.\n",
    "\n",
    "3. **Creating an Index:** Generates an index, which acts as a map or guide for quick retrieval of different genome parts. This index includes information about the positions of unique sequences in the genome.\n",
    "\n",
    "4. **Optimizing Search:** The created index speeds up the alignment process by allowing quick location of relevant genome regions, saving time especially when analyzing large datasets.\n",
    "\n",
    "In summary, `hisat2-build` creates an index from a reference genome, facilitating rapid alignment of sequence data during genetic analysis.\n",
    "\n",
    "#### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "87e5b03e-8e8e-42e5-bdb9-ae89a61757f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"hisat2_indexes/Otauri.*.ht2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Local offset rate: 3 (one in 8)\n",
      "  Local fTable chars: 6\n",
      "  Local sequence length: 57344\n",
      "  Local sequence overlap between two consecutive indexes: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  Data/O.tauri_genome.fna\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to read SNPs and splice sites: 00:00:00\n",
      "Using parameters --bmax 2441854 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2441854 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 1.6279e+06 (target: 2441853)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering GFM loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (2441854) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 1195229 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1195230 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (2441854) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 1294025 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1294026 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (2441854) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 1789994 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1789995 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (2441854) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1920036 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1920037 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (2441854) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1891077 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1891078 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (2441854) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 1964650 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1964651 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (2441854) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 1697345 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 1697346 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (2441854) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 1270858 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1270859 for bucket 8\n",
      "Exited GFM loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 2655580\n",
      "fchr[G]: 6510605\n",
      "fchr[T]: 10365900\n",
      "fchr[$]: 13023221\n",
      "Exiting GFM::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 8544067 bytes to primary GFM file: hisat2_indexes/Otauri.1.ht2\n",
      "Wrote 3255812 bytes to secondary GFM file: hisat2_indexes/Otauri.2.ht2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from GFM constructor\n",
      "Returning from initFromVector\n",
      "Wrote 5793245 bytes to primary GFM file: hisat2_indexes/Otauri.5.ht2\n",
      "Wrote 3312424 bytes to secondary GFM file: hisat2_indexes/Otauri.6.ht2\n",
      "Re-opening _in5 and _in5 as input streams\n",
      "Returning from HGFM constructor\n",
      "Headers:\n",
      "    len: 13023221\n",
      "    gbwtLen: 13023222\n",
      "    nodes: 13023222\n",
      "    sz: 3255806\n",
      "    gbwtSz: 3255806\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 0\n",
      "    eftabSz: 0\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 813952\n",
      "    offsSz: 3255808\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideGbwtSz: 48\n",
      "    sideGbwtLen: 192\n",
      "    numSides: 67830\n",
      "    numLines: 67830\n",
      "    gbwtTotLen: 4341120\n",
      "    gbwtTotSz: 4341120\n",
      "    reverse: 0\n",
      "    linearFM: Yes\n",
      "Total time for call to driver() for forward index: 00:00:05\n"
     ]
    }
   ],
   "source": [
    "mkdir -p hisat2_indexes\n",
    "hisat2-build Data/O.tauri_genome.fna hisat2_indexes/Otauri"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0e780f2-6c0c-454e-9563-4edc8ab4864f",
   "metadata": {},
   "source": [
    "## Check quality of the data\n",
    "\n",
    "### The tool used : `fastqc`\n",
    "\n",
    "The `fastqc` command is used in bioinformatics for quality control analysis of high-throughput sequencing data. Here's a brief overview of its functionality:\n",
    "\n",
    "1. **Quality Control Analysis:** FastQC analyzes the quality of sequencing data to identify potential issues or biases that may affect downstream analysis.\n",
    "\n",
    "2. **Input Data:** Takes input files in various formats (e.g., FASTQ) containing sequencing reads generated from experiments like RNA-seq, ChIP-seq, or whole-genome sequencing.\n",
    "\n",
    "3. **Analysis Modules:** FastQC performs various analysis modules to assess different aspects of data quality, including per-base sequence quality, per-sequence quality scores, sequence length distribution, GC content, sequence duplication levels, and overrepresented sequences.\n",
    "\n",
    "4. **Visual Reports:** Generates HTML reports containing summary statistics, graphs, and charts to visualize the quality metrics and identify any anomalies or patterns in the data.\n",
    "\n",
    "5. **Interpretation:** Users can interpret the FastQC reports to make informed decisions about data preprocessing steps, such as trimming adapters, filtering low-quality reads, or adjusting sequencing parameters.\n",
    "\n",
    "In summary, `fastqc` is a versatile tool for assessing the quality of sequencing data, providing valuable insights to researchers for optimizing experimental protocols and ensuring the reliability of downstream analyses.\n",
    "\n",
    "### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0595a1d5-68ac-4a3f-8b2e-cf0d7a3dfdde",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRR3099585_chr18.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Approx 5% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 10% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 15% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 20% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 25% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 30% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 35% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 40% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 45% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 50% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 55% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 60% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 65% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 70% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 75% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 80% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 85% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 90% complete for SRR3099585_chr18.fastq.gz\n",
      "Approx 95% complete for SRR3099585_chr18.fastq.gz\n",
      "Analysis complete for SRR3099585_chr18.fastq.gz\n",
      "Started analysis of SRR3099586_chr18.fastq.gz\n",
      "Approx 5% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 10% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 15% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 20% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 25% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 30% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 35% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 40% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 45% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 50% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 55% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 60% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 65% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 70% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 75% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 80% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 85% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 90% complete for SRR3099586_chr18.fastq.gz\n",
      "Approx 95% complete for SRR3099586_chr18.fastq.gz\n",
      "Analysis complete for SRR3099586_chr18.fastq.gz\n",
      "Started analysis of SRR3099587_chr18.fastq.gz\n",
      "Approx 5% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 10% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 15% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 20% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 25% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 30% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 35% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 40% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 45% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 50% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 55% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 60% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 65% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 70% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 75% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 80% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 85% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 90% complete for SRR3099587_chr18.fastq.gz\n",
      "Approx 95% complete for SRR3099587_chr18.fastq.gz\n",
      "Analysis complete for SRR3099587_chr18.fastq.gz\n",
      "Started analysis of SRR3105697_chr18.fastq.gz\n",
      "Approx 5% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 10% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 15% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 20% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 25% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 30% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 35% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 40% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 45% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 50% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 55% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 60% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 65% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 70% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 75% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 80% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 85% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 90% complete for SRR3105697_chr18.fastq.gz\n",
      "Approx 95% complete for SRR3105697_chr18.fastq.gz\n",
      "Analysis complete for SRR3105697_chr18.fastq.gz\n",
      "Started analysis of SRR3105698_chr18.fastq.gz\n",
      "Approx 5% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 10% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 15% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 20% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 25% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 30% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 35% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 40% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 45% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 50% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 55% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 60% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 65% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 70% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 75% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 80% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 85% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 90% complete for SRR3105698_chr18.fastq.gz\n",
      "Approx 95% complete for SRR3105698_chr18.fastq.gz\n",
      "Analysis complete for SRR3105698_chr18.fastq.gz\n",
      "Started analysis of SRR3105699_chr18.fastq.gz\n",
      "Approx 5% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 10% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 15% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 20% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 25% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 30% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 35% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 40% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 45% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 50% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 55% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 60% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 65% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 70% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 75% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 80% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 85% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 90% complete for SRR3105699_chr18.fastq.gz\n",
      "Approx 95% complete for SRR3105699_chr18.fastq.gz\n",
      "Analysis complete for SRR3105699_chr18.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "mkdir -p quality\n",
    "fastqc -o quality Data/*.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dcc9724-ffb3-44f9-8b95-5994d43a45a6",
   "metadata": {},
   "source": [
    "## Align data to reference genome\n",
    "\n",
    "### The tools we use : `hisat2` & `samtools`\n",
    "\n",
    "#### Hisat2\n",
    "\n",
    "`hisat2` is a bioinformatics tool commonly used for aligning high-throughput sequencing reads to a reference genome. Here's a concise overview of its functionality:\n",
    "\n",
    "1. **Alignment of Sequencing Reads:** Hisat2 aligns sequencing reads, typically obtained from RNA-seq, DNA-seq, or other sequencing experiments, to a reference genome or transcriptome.\n",
    "\n",
    "2. **Input Data:** Accepts input reads in FASTQ format, which contain short sequences of nucleotides generated by sequencing machines.\n",
    "\n",
    "3. **Reference Genome Alignment:** Aligns reads to a reference genome or transcriptome to determine their genomic origin or mapping position.\n",
    "\n",
    "4. **Sensitive and Fast Alignment:** Hisat2 employs advanced algorithms to achieve highly sensitive and accurate alignments while maintaining computational efficiency, making it suitable for analyzing large-scale sequencing datasets.\n",
    "\n",
    "5. **Spliced Alignment:** Capable of accurately aligning reads across splice junctions in eukaryotic genomes, facilitating gene expression analysis from RNA-seq data.\n",
    "\n",
    "6. **Output Formats:** Generates alignment results in standard formats such as SAM (Sequence Alignment/Map) or BAM (Binary Alignment/Map), which can be further processed or visualized using various bioinformatics tools.\n",
    "\n",
    "7. **Compatibility:** Hisat2 is compatible with downstream analysis tools and pipelines commonly used in bioinformatics, allowing seamless integration into data analysis workflows.\n",
    "\n",
    "In summary, `hisat2` is a versatile and efficient tool for aligning sequencing reads to reference genomes or transcriptomes, providing essential information for various genomic and transcriptomic analyses.\n",
    "\n",
    "#### samtools\n",
    "\n",
    "`Samtools` is a versatile bioinformatics tool widely used for manipulating and analyzing files in the SAM (Sequence Alignment/Map) and BAM (Binary Alignment/Map) formats, which are commonly used to represent sequence alignments generated by various sequencing platforms. Here's a concise overview of its functionality:\n",
    "\n",
    "1. **File Format Conversion:** Samtools provides utilities for converting between SAM and BAM formats, enabling efficient storage and manipulation of alignment data.\n",
    "\n",
    "2. **Alignment Inspection:** Allows users to view and extract alignment information from SAM and BAM files, including read sequences, mapping quality scores, alignment positions, and alignment flags.\n",
    "\n",
    "3. **Sorting and Indexing:** Provides functions to sort alignment files by genomic coordinates and create index files for rapid retrieval of specific genomic regions, improving the efficiency of downstream analysis tasks.\n",
    "\n",
    "4. **Alignment Filtering:** Enables users to filter alignments based on various criteria such as mapping quality, read length, alignment flags, and alignment position, allowing for the selection of high-quality alignments for further analysis.\n",
    "\n",
    "5. **Pileup Generation:** Computes pileup information, summarizing the depth and variation of aligned reads at each genomic position, which is useful for variant calling and other genomic analyses.\n",
    "\n",
    "6. **Variant Calling:** Supports variant calling from aligned reads, providing tools for identifying single nucleotide polymorphisms (SNPs), insertions, deletions, and structural variants in genomic data.\n",
    "\n",
    "7. **Integration with Workflows:** Samtools can be seamlessly integrated into bioinformatics workflows, serving as a fundamental component for processing and analyzing high-throughput sequencing data.\n",
    "\n",
    "In summary, `samtools` is a powerful toolset for manipulating and analyzing sequence alignment data, providing essential functionalities for a wide range of bioinformatics applications.\n",
    "\n",
    "### Code "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "29417932-345b-4157-8230-a233d7c846d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data/SRR3099585_chr18.fastq.gz\n",
      "313471 reads; of these:\n",
      "  313471 (100.00%) were unpaired; of these:\n",
      "    4896 (1.56%) aligned 0 times\n",
      "    237803 (75.86%) aligned exactly 1 time\n",
      "    70772 (22.58%) aligned >1 times\n",
      "98.44% overall alignment rate\n",
      "Data/SRR3099586_chr18.fastq.gz\n",
      "252818 reads; of these:\n",
      "  252818 (100.00%) were unpaired; of these:\n",
      "    6101 (2.41%) aligned 0 times\n",
      "    211303 (83.58%) aligned exactly 1 time\n",
      "    35414 (14.01%) aligned >1 times\n",
      "97.59% overall alignment rate\n",
      "Data/SRR3099587_chr18.fastq.gz\n",
      "275225 reads; of these:\n",
      "  275225 (100.00%) were unpaired; of these:\n",
      "    6829 (2.48%) aligned 0 times\n",
      "    236283 (85.85%) aligned exactly 1 time\n",
      "    32113 (11.67%) aligned >1 times\n",
      "97.52% overall alignment rate\n",
      "Data/SRR3105697_chr18.fastq.gz\n",
      "252211 reads; of these:\n",
      "  252211 (100.00%) were unpaired; of these:\n",
      "    5865 (2.33%) aligned 0 times\n",
      "    222825 (88.35%) aligned exactly 1 time\n",
      "    23521 (9.33%) aligned >1 times\n",
      "97.67% overall alignment rate\n",
      "Data/SRR3105698_chr18.fastq.gz\n",
      "405603 reads; of these:\n",
      "  405603 (100.00%) were unpaired; of these:\n",
      "    8490 (2.09%) aligned 0 times\n",
      "    350508 (86.42%) aligned exactly 1 time\n",
      "    46605 (11.49%) aligned >1 times\n",
      "97.91% overall alignment rate\n",
      "Data/SRR3105699_chr18.fastq.gz\n",
      "235166 reads; of these:\n",
      "  235166 (100.00%) were unpaired; of these:\n",
      "    8099 (3.44%) aligned 0 times\n",
      "    204267 (86.86%) aligned exactly 1 time\n",
      "    22800 (9.70%) aligned >1 times\n",
      "96.56% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "mkdir -p hisat2\n",
    "for fq in Data/*.fastq.gz ; do\n",
    "    echo ${fq} \n",
    "    libname=$(basename $fq .fastq.gz)\n",
    "    hisat2 -x hisat2_indexes/Otauri -q -U ${fq} -S hisat2/${libname}.sam\n",
    "    samtools view -b -o hisat2/${libname}.bam hisat2/${libname}.sam\n",
    "    samtools sort -o hisat2/${libname}-sort.bam hisat2/${libname}.bam\n",
    "    samtools index hisat2/${libname}-sort.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55582a57-1b53-4e42-ba55-d855ffb2fc24",
   "metadata": {},
   "source": [
    "## Count number of reads per gene\n",
    "\n",
    "### The tool we use : `htseq-count`\n",
    "\n",
    "The `htseq-count` tool is used in bioinformatics for quantifying gene expression from RNA sequencing (RNA-seq) data. Here's a brief overview of its functionality:\n",
    "\n",
    "1. **Counting Reads:** `htseq-count` counts the number of sequencing reads that align to features such as genes or exons in a reference genome.\n",
    "\n",
    "2. **Input Data:** Takes as input aligned reads in SAM (Sequence Alignment/Map) or BAM (Binary Alignment/Map) format, typically generated by aligners like HISAT2 or STAR.\n",
    "\n",
    "3. **Feature Annotation:** Requires a file containing annotations of genomic features, such as gene coordinates in GTF (Gene Transfer Format) or GFF (General Feature Format).\n",
    "\n",
    "4. **Counting Strategy:** Implements a strategy to assign aligned reads to features based on their alignment coordinates. Reads overlapping multiple features or features with overlapping coordinates can be handled according to specified preferences.\n",
    "\n",
    "5. **Output:** Produces a table or file containing the count of reads aligned to each feature, which can be used for downstream differential expression analysis or other gene expression studies.\n",
    "\n",
    "6. **Compatible with Downstream Tools:** The output of `htseq-count` is compatible with various statistical tools and workflows used for differential expression analysis, such as DESeq2 or edgeR.\n",
    "\n",
    "In summary, `htseq-count` is a valuable tool for quantifying gene expression from RNA-seq data, providing researchers with essential data for understanding the transcriptional landscape of biological samples.\n",
    "\n",
    "### Code\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "893ed5d6-19eb-4c68-aa5b-4e8585dd4d81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/shared/ifbstor1/home/ltroxler/My1rstnotebook\n",
      "bash: cd: My1rstnotebook: No such file or directory\n",
      "/shared/ifbstor1/home/ltroxler/My1rstnotebook\n"
     ]
    }
   ],
   "source": [
    "pwd\n",
    "cd My1rstnotebook\n",
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3366fc7d-250d-40fe-9ceb-2dda4aa957d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: htseq-count [options] alignment_file gff_file\n",
      "htseq-count: error: unrecognized arguments: /shared/home/ltroxler/My1rstnotebook/hisat2/SRR3099585_chr18-sort.bam /shared/home/ltroxler/My1rstnotebook/hisat2/SRR3099586_chr18-sort.bam /shared/home/ltroxler/My1rstnotebook/hisat2/SRR3099587_chr18-sort.bam /shared/home/ltroxler/My1rstnotebook/hisat2/SRR3105697_chr18-sort.bam /shared/home/ltroxler/My1rstnotebook/hisat2/SRR3105698_chr18-sort.bam /shared/home/ltroxler/My1rstnotebook/hisat2/SRR3105699_chr18-sort.bam /shared/home/ltroxler/My1rstnotebook/Data/O.tauri_annotation.gff\n"
     ]
    },
    {
     "ename": "",
     "evalue": "2",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "htseq-count -f bam r pos -s no -t gene -i ID -m intersection-nonempty /shared/home/ltroxler/My1rstnotebook/hisat2/*-sort.bam /shared/home/ltroxler/My1rstnotebook/Data/O.tauri_annotation.gff > counts.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e179ecef-2b30-4196-8c39-845494dd51ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ostta01g00010\t11\t10\t10\t1\t4\t7\n",
      "ostta01g00020\t3\t7\t10\t6\t7\t5\n",
      "ostta01g00030\t0\t1\t1\t2\t0\t2\n",
      "ostta01g00040\t20\t13\t17\t23\t16\t12\n",
      "ostta01g00050\t5\t18\t8\t12\t13\t11\n",
      "ostta01g00060\t4\t2\t4\t5\t2\t7\n",
      "ostta01g00070\t2\t3\t4\t1\t1\t2\n",
      "ostta01g00080\t2\t2\t0\t0\t0\t0\n",
      "ostta01g00090\t17\t81\t61\t45\t61\t65\n",
      "ostta01g00100\t11\t14\t10\t5\t19\t8\n"
     ]
    }
   ],
   "source": [
    "head counts.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd69e3cd-8207-4584-8c83-6f2820bda82a",
   "metadata": {},
   "source": [
    "## Differential analysis\n",
    "\n",
    "### The tool we use : `Desq2`\n",
    "\n",
    "DESeq2 is a software package for differential gene expression analysis from RNA sequencing (RNA-seq) data. It is widely used in molecular biology and genomics to identify genes whose expression varies significantly between different experimental conditions.\n",
    "\n",
    "### Code "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "338b4a17-93a6-4bdb-852c-2fe5b529cca6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook Deseq2.ipynb to script\n",
      "[NbConvertApp] Writing 795 bytes to Deseq2.r\n"
     ]
    }
   ],
   "source": [
    "jupyter nbconvert --to script Deseq2.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3b4d7ab6-118a-4bb0-ba86-ab35167ef31b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "R version 4.2.3 (2023-03-15) -- \"Shortstop Beagle\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(DESeq2)\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\u001b[?25h> \n",
      "> cts <- read.table(\"counts.txt\", sep=\"\\t\",row.names=1)\n",
      "\u001b[?25h> colnames(cts) = c(paste0(\"condition1_\",1:3),paste0(\"condition2_\",1:3))\n",
      "\u001b[?25h> head(cts)\n",
      "              condition1_1 condition1_2 condition1_3 condition2_1 condition2_2\n",
      "ostta01g00010           11           10           10            1            4\n",
      "ostta01g00020            3            7           10            6            7\n",
      "ostta01g00030            0            1            1            2            0\n",
      "ostta01g00040           20           13           17           23           16\n",
      "ostta01g00050            5           18            8           12           13\n",
      "ostta01g00060            4            2            4            5            2\n",
      "              condition2_3\n",
      "ostta01g00010            7\n",
      "ostta01g00020            5\n",
      "ostta01g00030            2\n",
      "ostta01g00040           12\n",
      "ostta01g00050           11\n",
      "ostta01g00060            7\n",
      "\u001b[?25h> \n",
      "> coldata = cbind.data.frame(condition= c(rep(\"condition1\", 3),rep(\"condition2\", 3)))\n",
      "\u001b[?25h> rownames(coldata) = c(paste0(\"condition1_\",1:3),paste0(\"condition2_\",1:3))\n",
      "\u001b[?25h> head(coldata)\n",
      "              condition\n",
      "condition1_1 condition1\n",
      "condition1_2 condition1\n",
      "condition1_3 condition1\n",
      "condition2_1 condition2\n",
      "condition2_2 condition2\n",
      "condition2_3 condition2\n",
      "\u001b[?25h> \n",
      "> dds <- DESeqDataSetFromMatrix(countData = cts,\n",
      "+                               colData = coldata ,\n",
      "+                               design= ~ condition)\n",
      "Warning message:\n",
      "In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "  some variables in design formula are characters, converting to factors\n",
      "\u001b[?25h> dds <- DESeq(dds)\n",
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "\u001b[?25h> resultsNames(dds) # lists the coefficients\n",
      "[1] \"Intercept\"                          \"condition_condition2_vs_condition1\"\n",
      "\u001b[?25h> \n",
      "> res <- results(dds, name=\"condition_condition2_vs_condition1\")\n",
      "\u001b[?25h> res\n",
      "log2 fold change (MLE): condition condition2 vs condition1 \n",
      "Wald test p-value: condition condition2 vs condition1 \n",
      "DataFrame with 7664 rows and 6 columns\n",
      "                        baseMean log2FoldChange     lfcSE      stat    pvalue\n",
      "                       <numeric>      <numeric> <numeric> <numeric> <numeric>\n",
      "ostta01g00010           7.088433     -1.4628099  0.813384 -1.798424 0.0721098\n",
      "ostta01g00020           6.075950     -0.2021194  0.806824 -0.250512 0.8021912\n",
      "ostta01g00030           0.936263      0.9463204  1.960157  0.482778 0.6292535\n",
      "ostta01g00040          16.498881     -0.0676998  0.572858 -0.118179 0.9059258\n",
      "ostta01g00050          10.719286      0.1597236  0.670239  0.238308 0.8116420\n",
      "...                          ...            ...       ...       ...       ...\n",
      "__no_feature            12184.10       0.214497  0.347472  0.617308  0.537031\n",
      "__ambiguous                 0.00             NA        NA        NA        NA\n",
      "__too_low_aQual             0.00             NA        NA        NA        NA\n",
      "__not_aligned            6489.58       0.261774  0.341437  0.766682  0.443271\n",
      "__alignment_not_unique  38609.06      -0.696942  0.467462 -1.490905  0.135986\n",
      "                            padj\n",
      "                       <numeric>\n",
      "ostta01g00010                 NA\n",
      "ostta01g00020                 NA\n",
      "ostta01g00030                 NA\n",
      "ostta01g00040                 NA\n",
      "ostta01g00050                 NA\n",
      "...                          ...\n",
      "__no_feature            0.850693\n",
      "__ambiguous                   NA\n",
      "__too_low_aQual               NA\n",
      "__not_aligned           0.816958\n",
      "__alignment_not_unique  0.584435\n",
      "\u001b[?25h> \n",
      "> plotMA(res, ylim=c(-2,2))\n",
      "\u001b[?25h> \n",
      "> png(file = \"deseq2_demo.png\", width = 800, height = 700)\n",
      "\u001b[?25h> plotMA(res, ylim=c(-2,2))\n",
      "\u001b[?25h> dev.off()\n",
      "pdf \n",
      "  2 \n",
      "\u001b[?25h> \n",
      "> write.table(res,\"deseq2_demo.tsv\", quote=F, sep=\"\\t\", col.names=T)\n",
      "\u001b[?25h> \n",
      "> \n",
      "> \n",
      "\u001b[?25h\n"
     ]
    }
   ],
   "source": [
    "R < Deseq2.r --no-save  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1072ab84-4d3e-4154-9663-1191f9a1a889",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
